contribs/gnostats/agent/collector.go

```diff
@@ -0,0 +1,156 @@
+	Send(context.Context) ([]any, error)
+}
+
+// CollectDynamic collects dynamic info from the Gno node using RPC
+func (c *collector) CollectDynamic(ctx context.Context) (*proto.DynamicInfo, error) {
+	// Create a new batch of RPC requests
+	batch := c.caller.NewBatch()
+
+	for _, request := range [](func() error){
+		// Request Status to get address, moniker and validator info
+		batch.Status,
+		// Request NetInfo to get peers info
+		batch.NetInfo,
+		// Request NumUnconfirmedTxs to get pending txs
+		batch.NumUnconfirmedTxs,
+		// Request Block to get the last block timestamp and proposer
+		func() error { return batch.Block(nil) },
+		// Request BlockResults to get the last block number, gas used and gas wanted
+		func() error { return batch.BlockResults(nil) },
+	} {
+		if err := request(); err != nil {
+			return nil, fmt.Errorf("unable to batch request: %w", err)
+		}
+	}
+
+	// Send the batch of requests
+	results, err := batch.Send(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to send batch: %w", err)
+	}
+
+	// Cast responses to the appropriate types
+	var (
+		status  = results[0].(*ctypes.ResultStatus)
+		netInfo = results[1].(*ctypes.ResultNetInfo)
+		uncTxs  = results[2].(*ctypes.ResultUnconfirmedTxs)
+		blk     = results[3].(*ctypes.ResultBlock)
+		blkRes  = results[4].(*ctypes.ResultBlockResults)
+	)
+
+	// Convert the list of peers from NetInfo to proto type
+	peers := make([]*proto.PeerInfo, len(netInfo.Peers))
+	for i, peer := range netInfo.Peers {
+		peers[i] = &proto.PeerInfo{
+			Moniker: peer.NodeInfo.Moniker,
+		}
+		if peer.NodeInfo.NetAddress != nil {
+			peers[i].P2PAddress = peer.NodeInfo.NetAddress.String()
+		}
+	}
+
+	// Determine if the node is a validator using validatorInfo
+	// @TODO: Check if this is right with @zivkovicmilos
+	isValidator := status.ValidatorInfo.Address.ID().String() != ""
+
+	// Get gas used / wanted in DeliverTxs (if any)
+	var gasUsed, gasWanted uint64
+	if blkRes.Results != nil && len(blkRes.Results.DeliverTxs) > 0 {
+		gasUsed = uint64(blkRes.Results.DeliverTxs[0].GasUsed)
+		gasWanted = uint64(blkRes.Results.DeliverTxs[0].GasWanted)
+	}
+
+	// Fill the DynamicInfo fields with the corresponding values
+	return &proto.DynamicInfo{
+		Address:     status.NodeInfo.ID().String(),
+		Moniker:     status.NodeInfo.Moniker,
+		IsValidator: isValidator,
+		NetInfo: &proto.NetInfo{
+			P2PAddress: status.NodeInfo.NetAddress.String(),
+			Peers:      peers,
+		},
+		PendingTxs: uint64(uncTxs.Total),
+		BlockInfo: &proto.BlockInfo{
+			Number:    uint64(blk.Block.Height),
+			Timestamp: uint64(blk.Block.Time.Unix()),
+			GasUsed:   gasUsed,
+			GasWanted: gasWanted,
+			Proposer:  blk.Block.ProposerAddress.ID().String(),
+		},
+	}, nil
+}
+
+// CollectStatic collects static info on the Gno node using RPC
+func (c *collector) CollectStatic(ctx context.Context) (*proto.StaticInfo, error) {
+	// Use a batch instead of a single request to allow passing a context
+	batch := c.caller.NewBatch()
```

Comment:
You don't a batch for a single request, you can just call it flat-out

Reply:
Calling it directly also gives you type safety

Reply:
I did this to have the option to pass a context to `Send` (in case of a cancel/timeout).
Would you prefer that I make a simple call without context anyway?

Reply:
You can init the client with this option, so you can control the request:
```go
// WithRequestTimeout sets the request timeout
func WithRequestTimeout(timeout time.Duration) Option {
	return func(client *RPCClient) {
		client.requestTimeout = timeout
	}
}
```

In retrospect, these client methods (rpc) should take in a context, since we spawn a context anyways when handling them (`context.Background()`), but this is something for a future PR. @gfanton what do you think?

cc @Kouteki for the issue

Reply:
The `WithRequestTimeout` will not work if the caller cancels the context. Should I still make a change using this option? Or would you prefer that I wait for the context parameter to be added at the RPC client level? Or I could also leave the code as is with a `TODO` comment? What do you think?

Reply:
Let's roll without it, I don't see it as a blocker